src/openms/source/QC/Contaminants.cpp

```diff
@@ -0,0 +1,220 @@
+      {
+        // peptideidentifications in feature f is not empty
+        if (id.getHits().empty())
+        {
+          ++ feature_has_no_sequence;
+          continue;
+        }
+
+        else
+        {
+          // the one existing peptideidentification has atleast one getHits entry
+          auto &pep_hit = f.getPeptideIdentifications()[0].getHits()[0];
+          String key = (pep_hit.getSequence().toUnmodifiedString());
+          this->compare_(key, f, total, cont, sum_total, sum_cont);
+        }
+      }
+    }
+    //save the contaminants ratio in object before searching through the unassigned peptideidentifications
+    ContaminantsSummary final;
+    final.assigned_contaminants_ratio = (cont / double(total));
+
+    final.empty_features.first = feature_has_no_sequence;
+    final.empty_features.second = features.size();
+
+    UInt64 utotal = 0;
+    UInt64 ucont = 0;
+
+    //Change the assigned contaminants ratio to total contaminants ratio by adding the unassigned.
+    //Additionally save the unassigned contaminants ratio and add the is_contaminant = 0/1 to the first hit of the unassigned peptideidentifications.
+    //if (features.getUnassignedPeptideIdentifications().empty())
+    //{
+    //  return;
+    //}
+    for (auto& fu : features.getUnassignedPeptideIdentifications())
+    {
+      if ( fu.getHits().empty()) continue;
+      auto& fu_hit = fu.getHits()[0];
+      String key = (fu_hit.getSequence().toUnmodifiedString());
+      ++utotal;
+
+      //peptide is not in contaminant database
+      if (!digested_db_.count(key))
+      {
+        fu.getHits()[0].setMetaValue("is_contaminant", 0);
+        continue;
+      }
+
+      //peptide is contaminant
+      ++ucont;
+      fu.getHits()[0].setMetaValue("is_contaminant", 1);
+    }
+    total += utotal;
+    cont += ucont;
+
+    //save all ratios and the intensity to the object
+    final.all_contaminants_ratio = (cont / double(total));
+    final.unassigned_contaminants_ratio = (ucont / double(utotal));
+    final.assigned_contaminants_intensity_ratio = (sum_cont / sum_total);
+
+
+    //add the object to the results vector
+    results_.push_back(final); //save all ratios in results_ as tuples
+  }
+
+  const std::vector<Contaminants::ContaminantsSummary>& Contaminants::getResults()
+  {
+    return results_;
+  }
+
+  //Check if peptide is in contaminants database or not and add the is_contaminant = 0/1.
+  // If so, raise the contaminant ratio.
+  void Contaminants::compare_(const String& key,
+                             Feature& f,
+                             Int64& total,
+                             Int64& cont,
+                             double& sum_total,
+                             double& sum_cont)
+  {
+    ++total;
+    sum_total += f.getIntensity();
+    //peptide is not in contaminant database
+    if (!digested_db_.count(key))
+    {
+      f.getPeptideIdentifications()[0].getHits()[0].setMetaValue("is_contaminant", 0);
```

Comment:
potentially buggy if feature has 2 or more PepIds (since you annotate the first, always). Pass in a PeptideHit instead of a Feature and annotate that. Also extend the tests to check this is correct